File: game.rb
require 'securerandom'

class Game
  def initialize(moves)
    @moves = moves
    @secret_key = SecureRandom.random_bytes(32) # Generate a 256-bit secret key
  end

  def start
    computer_move_index = rand(@moves.length)
    computer_move = @moves[computer_move_index]

    hmac = calculate_hmac(computer_move)

    puts "HMAC: #{hmac}\n"
    display_menu

    print "Enter your move: "
    user_input = gets.chomp

    if user_input == "?"
      display_help_table
      start
      return
    end

    parsed_choice = user_input.to_i
    if parsed_choice.to_s != user_input || parsed_choice < 0 || parsed_choice > @moves.length
      puts 'Invalid input. Please enter a valid move number or "?" for help.'
      start
      return
    end

    if parsed_choice == 0
      puts "Exiting the game."
      return
    end

    user_move = @moves[parsed_choice - 1]
    puts "Your move: #{user_move}"
    puts "Computer move: #{computer_move}"
    result = evaluate_result(user_move, computer_move)
    puts result
    puts "HMAC key: #{@secret_key.unpack('H*').first}"
  end

  private

  def display_menu
    puts "Available moves:"
    @moves.each_with_index { |move, index| puts "#{index + 1} - #{move}" }
    puts "0 - exit"
    puts "? - help"
    puts '\nEnter your move or "?" for help:'
  end

  def display_help_table
    puts "\nResults from the user's point of view:"
    # Prepare header row with emphasized style
    header_row = "+--------------+#{'--------+'.repeat(@moves.length)}"
    header_row_with_emphasis = "\e[1m#{header_row}\e[0m"

    puts header_row_with_emphasis
    # Prepare header with PC vs. User
    header = "| v PC\\User >  |"
    @moves.each { |move| header += " #{move.ljust(6)} |" }
    puts header
    puts header_row_with_emphasis
    # Render the table body
    @moves.each do |move1|
      row = "| #{move1.ljust(12)} |"
      @moves.each do |move2|
        row += " #{evaluate_result(move1, move2).ljust(7)}|"
      end
      puts row
      puts header_row_with_emphasis
    end
  end

  def calculate_hmac(message)
    hmac = OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha256'), @secret_key, message)
    hmac.unpack('H*').first
  end

  def evaluate_result(user_move, computer_move)
    user_index = @moves.index(user_move)
    computer_index = @moves.index(computer_move)
    number_of_moves = @moves.length

    # Calculate the winning moves for the user
    winning_moves = []
    (1..number_of_moves / 2).each do |i|
      winning_moves.push((user_index + i) % number_of_moves)
    end

    return "Draw" if user_index == computer_index
    return "Win" if winning_moves.include?(computer_index)
    return "Lose"
  end
end

# Extract moves from command-line arguments, skipping the first two arguments (ruby executable and script filename)
moves = ARGV
if moves.length < 3 || moves.length.even?
  puts "Invalid number of moves. Please provide an odd number of moves greater than 2."
else
  game = Game.new(moves)
  game.start
end

File: game.js
const crypto = require("crypto");
const readline = require("readline");

class Game {
  constructor(moves) {
    this.moves = moves;
    this.secretKey = crypto.randomBytes(32); // Generate a 256-bit secret key
    this.interface = readline.createInterface(process.stdin, process.stdout);
  }

  start() {
    const computerMoveIndex = Math.floor(Math.random() * this.moves.length);
    const computerMove = this.moves[computerMoveIndex];

    const hmac = this.calculateHMAC(computerMove);

    console.log(`HMAC: ${hmac}\n`);
    this.displayMenu();

    this.interface.question("Enter your move: ", (userInput) => {
      const userChoice = userInput.trim();
      if (userChoice === "?") {
        this.displayHelpTable();
        this.start();
        return;
      }

      const parsedChoice = parseInt(userChoice);
      if (
        isNaN(parsedChoice) ||
        parsedChoice < 0 ||
        parsedChoice > this.moves.length
      ) {
        console.log(
          'Invalid input. Please enter a valid move number or "?" for help.'
        );
        this.start();
        return;
      }

      if (parsedChoice === 0) {
        console.log("Exiting the game.");
        this.interface.close();
        return;
      }

      const userMove = this.moves[parsedChoice - 1];
      console.log(`Your move: ${userMove}`);
      console.log(`Computer move: ${computerMove}`);
      const result = this.evaluateResult(userMove, computerMove);
      console.log(result);
      console.log(`HMAC key: ${this.secretKey.toString("hex")}`);
      this.interface.close();
    });
  }

  displayMenu() {
    console.log("Available moves:");
    this.moves.forEach((move, index) => {
      console.log(`${index + 1} - ${move}`);
    });
    console.log("0 - exit");
    console.log("? - help");
    console.log('\nEnter your move or "?" for help:');
  }

  displayHelpTable() {
    console.log("\nResults from the user's point of view:");
    // Prepare header row with emphasized style
    const headerRow = `+--------------+${"--------+".repeat(
      this.moves.length
    )}`;
    let headerRowWithEmphasis = "\x1b[1m" + headerRow + "\x1b[0m";

    console.log(headerRowWithEmphasis);
    // Prepare header with PC vs. User
    let header = "| v PC\\User >  |";
    this.moves.forEach((move) => {
      header += ` ${move.padEnd(6)} |`;
    });
    console.log(header);
    console.log(headerRowWithEmphasis);
    // Render the table body
    for (let i = 0; i < this.moves.length; i++) {
      let row = `| ${this.moves[i].padEnd(12)} |`;
      for (let j = 0; j < this.moves.length; j++) {
        row += ` ${this.evaluateResult(this.moves[i], this.moves[j]).padEnd(
          7
        )}|`;
      }
      console.log(row);
      console.log(headerRowWithEmphasis);
    }
  }

  calculateHMAC(message) {
    const hmac = crypto.createHmac("sha256", this.secretKey);
    hmac.update(message);
    return hmac.digest("hex");
  }

  evaluateResult(userMove, computerMove) {
    const userIndex = this.moves.indexOf(userMove);
    const computerIndex = this.moves.indexOf(computerMove);
    const numberOfMoves = this.moves.length;

    // Calculate the winning moves for the user
    const winningMoves = [];
    for (let i = 1; i <= Math.floor(numberOfMoves / 2); i++) {
        winningMoves.push((userIndex + i) % numberOfMoves);
    }

    if (userIndex === computerIndex) {
        return "Draw";
    } else if (winningMoves.includes(computerIndex)) {
        return "Win";
    } else {
        return "Lose";
    }
}

}

// Extract moves from command-line arguments, skipping the first two arguments (node executable and script filename)
const moves = process.argv.slice(2);
if (moves.length < 3 || moves.length % 2 === 0) {
  console.log(
    "Invalid number of moves. Please provide an odd number of moves greater than 2."
  );
} else {
  const game = new Game(moves);
  game.start();
}
